src/main/java/com/pingcap/tikv/operation/DAGIterator.java

```diff
@@ -0,0 +1,164 @@
+import com.pingcap.tikv.region.TiRegion;
+import com.pingcap.tikv.row.Row;
+import com.pingcap.tikv.row.RowReader;
+import com.pingcap.tikv.row.RowReaderFactory;
+import com.pingcap.tikv.types.DataType;
+import com.pingcap.tikv.types.DataTypeFactory;
+import com.pingcap.tikv.types.Types;
+import com.pingcap.tikv.util.RangeSplitter;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Objects;
+
+public class DAGIterator implements Iterator<Row> {
+  protected final TiSession session;
+  private final List<RangeSplitter.RegionTask> regionTasks;
+  private final boolean indexScan;
+  private TiDAGRequest dagRequest;
+  private static final DataType[] handleTypes =
+          new DataType[]{DataTypeFactory.of(Types.TYPE_LONG)};
+  private RowReader rowReader;
+  private CodecDataInput dataInput;
+  private boolean eof = false;
+  private int taskIndex;
+  private int chunkIndex;
+  private List<Chunk> chunkList;
+  private SchemaInfer schemaInfer;
+
+  public DAGIterator(TiDAGRequest req,
+                     List<RangeSplitter.RegionTask> regionTasks,
+                     TiSession session,
+                     boolean indexScan) {
+    this.dagRequest = req;
+    this.session = session;
+    this.regionTasks = regionTasks;
+    this.indexScan = indexScan;
+    this.schemaInfer = SchemaInfer.create(req);
+  }
+
+  public DAGIterator(TiDAGRequest req, TiSession session, RegionManager rm,
+                        boolean indexScan) {
+    this(req, RangeSplitter.newSplitter(rm).splitRangeByRegion(req.getRanges()), session, indexScan);
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (eof) {
+      return false;
+    }
+
+    while (null == chunkList ||
+            chunkIndex >= chunkList.size() ||
+            dataInput.available() <= 0
+            ) {
+      // First we check if our chunk list has remaining chunk
+      if (tryAdvanceChunkIndex()) {
+        createDataInputReader();
+      }
+      // If not, check next region
+      else if (!readNextRegionChunks()) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  @Override
+  public Row next() {
+    if (hasNext()) {
+      if (indexScan) {
+        return rowReader.readRow(handleTypes);
+      } else {
+        return rowReader.readRow(this.schemaInfer.getTypes().toArray(new DataType[0]));
+      }
+    } else {
+      throw new NoSuchElementException();
+    }
+  }
+
+  private boolean tryAdvanceChunkIndex() {
+    if (null == chunkList || chunkIndex >= chunkList.size() - 1) {
+      return false;
+    }
+
+    chunkIndex++;
+    return true;
+  }
+
+  private boolean readNextRegionChunks() {
+    if (null == regionTasks || taskIndex >= regionTasks.size()) {
+      return false;
+    }
+
+    RangeSplitter.RegionTask regionTask = regionTasks.get(taskIndex++);
+    List<Chunk> chunks = createClientAndSendReq(regionTask, this.dagRequest);
+    if (null == chunks || chunks.isEmpty()) {
```

Comment:
It is legal, but is it better you can do `chunks == null`.  This makes more sense which means it checking chunks is null or not.

Reply:
This is probably a habit learned from `C`, if , I will change it to a more meaningful `obj == null` 
 `Java` version.

Reply:
Done.